### ISSUE TITLE
Router collapses to p_skip ≡ 1 in stage-2 training — task-loss gradient to the skipping controllers appears to vanish

### ISSUE THREAD

**Issue by MingCS907** [PKU-SEC-Lab/DYSL_VLA#2]
**Environment**
Model OpenVLA-oft 7B, LIBERO-10 (libero_10_no_noops)
Config released train.sh: lambda1=1e-2, lambda2=1, adapter_epoch=8, router_epoch=200, static_layers='0,1,4,8,12,15,18,22,24,26,29,30,31,32', medium_feature=1024, max_steps=350005
Hardware 4090 batch_size=2, grad_accumulation_steps=16
Training actually completed 192k（-150k）

**Summary**
After the two-stage procedure converges, every skipping controller (RouterModule) outputs p_skip ≈ 1.00, i.e. the router degenerates into a constant "always skip" predictor. The number of executed layers becomes pinned to the static-layer set, and the adaptive behaviour shown in Fig. 4(a), Fig. 6 and Fig. 8 of the paper is not reproducible with the released code and config.
We would like to confirm whether we are missing a training detail, or whether this is a known limitation of the two-stage objective.

**What we observe**
p_skip per dynamic layer vs. training step: <figure / numbers> — all controllers converge to ≈ 1.0 and stay there.
Distribution of executed layers per inference: <histogram>. With static_layers='0,1,4,8,12,15,18,22,24,26,29,30,31,32' the static set has 13 of 32 layers; we measure <13 + k> layers.
Latency (<GPU>): <measured ms> vs. the paper's 27.4 ms (A6000) / 345 ms (Jetson Orin).
Note that p_skip ≈ 1 (rather than ≈ 0.5, the value at random initialisation) indicates the router phase did run and was actively driven to saturation — this is not a "router never trained" situation.
Step budget at which we measured the saturation: the router phase had received only ≈ <4k>–<12k> gradient steps (our checkpoints sit at absolute steps <162k> / <170k>, i.e. ≈ <4k> / <12k> steps after the switch at ≈ 157.7k), whereas a full run would give the router ≈ 192k steps. We therefore cannot yet exclude that the router would be pulled back later — although the analysis below says the task-loss path to it is already ≈ 0.

**Why we think this happens**
1. Stage 1 removes the task-loss signal to the router.

Stage 1 (Eq. 4, finetune.py:1260–1266) trains each adapter against the true activation at the next static layer:

adapter_loss += F.mse_loss(adapter_results[i], adapter_results[cfg.static_layers[layer_id]])
So after stage 1, adapter_i(x_i) ≈ L_{s_i−1}(…L_i(x_i)) on the training distribution.

2. Stage 2 mixes exactly those two now-equal branches.

The stage-2 forward (Eq. 5, modeling_llama.py:1212) is

hidden_states = skip_problity[0] * x_skip + skip_problity[1] * hidden_states
with (p₀, p₁) = F.gumbel_softmax(logits, tau=1, hard=True) in training (modeling_llama.py:115). The task loss therefore reaches the router only through

∂L_task/∂p₀  ∝  ( x_skip − x_computed ) · ∂L_task/∂x
and x_skip − x_computed is exactly the residual that stage 1 minimised. As stage 1 succeeds, the task-loss path to the router vanishes — for both the soft relaxation and the straight-through estimator.

3. **The only remaining term rewards skipping, unconditionally.**

Eq. 6 (finetune.py:1255) adds

router_loss += skip_problitys[tmp][1] * (cfg.static_layers[i] - exit_id[i])
i.e. p₁ · (s_i − i) = P(no-skip) × (layers that would be skipped). This term is monotonically decreasing in p₁, so with λ₁ = 1e-2 and no opposing term, the optimum is p₁ → 0, i.e. p_skip → 1 — unbounded saturation.

4. Consequence: "always skip" is arguably the optimum of the released objective, not an optimisation accident.

Once Eq. 4 has converged, the adapter is (on the training distribution, in activation-MSE terms) a good substitute for the skipped block, so skipping does not measurably increase the task loss. Under the released objective there is nothing that makes skipping cost anything at the steps where the substitution is wrong. The paper's own motivation — a few critical actions dominate task success (Fig. 1) — is not represented anywhere in stage 2's loss.

5. A testable prediction. If p_skip ≡ 1, the executed path should be exactly static layers ∪ {dynamic layers before the allow point}. Since allow_point is initialised at each segment's first dynamic layer (modeling_llama.py:1049–1056), the router should contribute nothing and the kept-layer count should equal |static| whenever the allow points do not move. Our measurements are consistent with this — <confirm / numbers>.

**Comment by mira687** [comment]
Your gradient argument holds up when I read the code against it, and two things in the released implementation make it worse than the issue states.

**The adapters are never frozen at the stage switch.** `finetune.py:968-979` sets `requires_grad_(True)` on both `router` and `layer_skip`; `trainable_params` and `AdamW` are built at `:1047-1050`; the loop starts at `:1153`. Nothing re-freezes anything when the stage-2 condition at `:1158` flips. So `layer_skip` keeps training in stage 2 — but Eq. 4's MSE anchor is gone, and the only gradient left for it is the task loss, which reaches it precisely on the steps the router skipped (`hard=True` makes the coefficient on `x_skip` at `:1212` exactly 1 or 0). That closes your loop: skip → adapter gets task-trained → `x_skip − x_computed` shrinks further → even less task-loss pressure on the router. Stage 1 doesn't remove the signal once; stage 2 keeps removing it.

**`:1188` takes batch element 0 only.** `self.router[b_idx](hidden_states)[0]` turns the `(bsz, 2)` output into a `(2,)`, and that one sample's decision gates the whole batch at `:1212`. Running the released `RouterModule` on CPU at `bsz=2`, the per-row gradient norms of `p1` from row 0 are `[0.327, 0.0]` w.r.t. the logits and `[0.139, 0.0]` w.r.t. the per-sample inputs — row 1 neither influences the decision nor receives gradient. `skip_problitys` inherits the same slice, so the penalty term is sample-0's too. Inference takes `[0][0]` at bsz=1, i.e. genuinely per-sample. At `batch_size=2`, half your samples are routed by another sample's content during training, which is a problem for the per-input adaptivity in Fig. 4(a) independently of the saturation.

On scale: under your own `exit_id` sampler (weights `1/(j−a+1)`) with `static_layers='0,1,4,...'`, the expected sum of `(static_layers[i] − exit_id[i])` across the 9 active segments is **15.76**, so at λ1=1e-2 the skip reward is ≈0.158 whenever every router computes — with a constant gradient, against a task-loss path of ≈0. The `print` at `:1258` already logs `loss` next to the sampled probabilities, so both sides of that comparison are in the logs you already have.

No GPU here, so this is your code read against a CPU reproduction of `RouterModule`, not a training run.
